javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll

```diff
@@ -168,8 +212,221 @@ private module Mongoose {
   /**
    * A Mongoose collection object.
    */
-  class Model extends MongoDB::Collection {
+  class Model extends DataFlow::SourceNode {
     Model() { this = getAMongooseInstance().getAMemberCall("model") }
+
+    private DataFlow::SourceNode ref(DataFlow::TypeTracker t) {
+      result = this and
+      t.start()
+      or
+      exists(DataFlow::TypeTracker t2 | result = ref(t2).track(t2, t))
+    }
+
+    DataFlow::SourceNode ref() { result = ref(DataFlow::TypeTracker::end()) }
+  }
+
+  /**
+   * Provides signatures for the Model methods.
+   */
+  module ModelMethodSignatures {
+    /**
+     * Holds if Model method `name` interprets parameter `n` as a query.
+     */
+    predicate interpretsArgumentAsQuery(string name, int n) {
+      // implement lots of the MongoDB collection interface
+      MongoDB::CollectionMethodSignatures::interpretsArgumentAsQuery(name, n)
+      or
+      name = "findByIdAndUpdate" and n = 1
```

Comment:
AFAIU the second argument describes how to modify the value(s) selected by the filter. If the attacker doesn't also control the first argument, does this in any way grant them access to other parts of the database?

I think we should be a consider exactly what constitutes a NoSQL injection, because inserting user-controlled values in specific rows of a database isn't usually enough by itself.

In many of the results, this is the sink, but notably the attacker also controls the first argument. So they're TPs, but with proper authz checks on the ID, and [validation](https://mongoosejs.com/docs/validation.html) on the Schema behind the model, I'm not so sure.

Reply:
The `update-query` has a lot of expressive power AFAIU, see this example:

https://docs.mongodb.com/manual/reference/method/db.collection.update/#update-method-examples
```
   [
     { $set: { average : { $trunc: [ { $avg: "$tests" }, 0 ] }, lastUpdate: "$$NOW" } },
     { $set: { grade: { $switch: {
                           branches: [
                               { case: { $gte: [ "$average", 90 ] }, then: "A" },
                               { case: { $gte: [ "$average", 80 ] }, then: "B" },
                               { case: { $gte: [ "$average", 70 ] }, then: "C" },
                               { case: { $gte: [ "$average", 60 ] }, then: "D" }
                           ],
                           default: "F"
     } } } }
   ]
```

So from this you could either leak secret data (perhaps restricted to the document that is being updated), or perform a DOS attack using expensive operators.

(Besides, we already had those `update-query` sinks in the old version (generally present at the `index = 1` lines).)